dbms/src/Flash/Mpp/MPPTaskManager.cpp

```diff
@@ -31,11 +34,46 @@ extern const char random_task_manager_find_task_failure_failpoint[];
 extern const char pause_before_register_non_root_mpp_task[];
 } // namespace FailPoints
 
+namespace
+{
+// Print log for MPPTask which hasn't been removed for over 5 minutes.
+std::vector<std::pair<String, double>> getLongLiveMPPTasks(const std::unordered_map<String, Stopwatch> & monitored_tasks)
+{
+    std::vector<std::pair<String, double>> long_live_tasks;
+    for (const auto & iter : monitored_tasks)
+    {
+        auto alive_time = iter.second.elapsedSeconds();
+        if (alive_time >= 300)
+            long_live_tasks.push_back(std::make_pair(iter.first, alive_time));
+    }
+
+    return long_live_tasks;
+}
+} // namespace
+
 MPPTaskManager::MPPTaskManager(MPPTaskSchedulerPtr scheduler_)
     : scheduler(std::move(scheduler_))
     , log(Logger::get())
+    , is_shutdown(false)
 {}
 
+MPPTaskManager::~MPPTaskManager()
+{
+    std::lock_guard lock(mu);
+    is_shutdown = true;
+    monitor_cv.notify_all();
+}
+
+void MPPTaskManager::startMonitorMPPTaskThread()
```

Comment:
I suggest to put this monitor thread to `BackgroundSerive` in TMTContext, and add some sync method to make sure the monitor thread always visit a valid mpp task manager.